### DISCUSSION TITLE
props to attach functions to events

### DISCUSSION THREAD

**Discussion by vitordino** [ariakit/ariakit#412] [Questions] [answered]
### Motivation

I’m refactoring a codebase to use `reakit`’s component underneath, and I keep missing some props that are common on controlled components (eg.: form inputs)..

as we have a `visible` "value" prop both on state and components (eg.: [`useHiddenState`](https://github.com/reakit/reakit/blob/master/packages/reakit/src/Hidden/HiddenState.ts#L83) and used on [`<Hidden/>`](https://github.com/reakit/reakit/blob/master/packages/reakit/src/Hidden/Hidden.ts#L13) itself), in my opinion, we should have a declarative way to attach other functions (side effects) to its events — something as `onShow`, `onHide` and `onToggle` props.

### Example

```
const state = useDialogState({
  onShow: () => console.log('i’m showing'),
  onHide: () => console.log('i’m hiding'),
  onToggle: e => console.log(`i’m ${e.visible ? 'showing' : 'hiding'}`),
})
```

### Possible implementations

this feature probably would be implemented by first plugging some props to [`useHiddenState`](https://github.com/reakit/reakit/blob/master/packages/reakit/src/Hidden/HiddenState.ts) then, if necessary, for each `state` hook that implements new functions.

[a comment by an automated account is not shown]

**Comment by vitordino** [comment]
currently i’m dealing with a custom hook on top of `reakit`’s hooks, and it looks like this:

```
const noop = () => ({})

export const useDialogStateWithAttachedEvents = ({
  onShow = noop,
  onHide = noop,
  onToggle = noop,
  ...props
}) => {
  const state = useDialogState(props)

  const show = e => {
    state.show(e)
    onShow(e)
  }

  const hide = e => {
    state.hide(e)
    onHide(e)
  }

  const toggle = e => {
    state.toggle(e)
    onToggle(e)
  }

  return { ...state, show, hide, toggle }
}
```

**Comment by diegohaz** [comment, marked as answer]
I usually suggest this approach:
```jsx
const state = useDialogState();

React.useEffect(() => {
  if (state.visible) {
    // onShow
  } else {
    // onHide
  }
}, [state.visible]);
```
Related: https://spectrum.chat/reakit/general/how-to-react-on-dialog-events~710e0748-288e-4196-a030-2243eede50ca

You can also create something like [`useUpdateEffect`](https://github.com/reakit/reakit/blob/e4fadc07930c426b564a00700c3d9a44d34c80ff/packages/reakit-utils/src/useUpdateEffect.ts) to handle only updates.

But, yes, wapping it within an enhanced hook is also a good option.

**Comment by vitordino** [comment]
nice @diegohaz thanks for your fast reply, it really helped me 🙌 ..

just to be clear, we don’t plan on putting these props on the official API, right?
if so, i think we may tag this as a question and close it

**Comment by diegohaz** [comment]
Yes, right now I don't think it makes sense including this in the API as this is easily doable in the user land. But I'm glad that you opened the issue so more people can find it if they have a similar need. :)
